### ISSUE TITLE
Recent 0.32 release breaks Python 3.4 / Ubuntu Trusty (rollback to v0.31.1)

### ISSUE THREAD

**Issue by ryanellis** [pypa/wheel#257]
Using PBuilder, I'm getting this error on at least three dependencies:
```AttributeError: 'Requirement' object has no attribute 'url'```
Causing the wheel to not build.

This is happening on Python v3.4 (OS: Ubuntu/Trusty)

Interim solution: use `wheel==0.31.1` when installing/pinning

FYI: the third-party projects that were failing:
- Pillow
- celery-flower
- tornado

**Comment by benoit-pierre** [comment]
What's the full traceback?

**Comment by benoit-pierre** [comment]
Also, what version of setuptools / packaging do you have installed? And are those provided by the distribution?

**Comment by robvdl** [comment]
We are building with a pbuilder ubuntu 14.04 environment so basically it's a clean ubuntu 14.04 installation.

During the build we are creating a new python3 virtualenv using python 3.4 from trusty and then upgrading pip and wheel to the latest.

We then build wheels for our dependencies and app itself, during the building of our app wheel it fails on one of the dependencies with that AttributeError. This has started happening with wheel v0.32.0 released recently.

By pinning wheel at 0.31.x our build is working fine again. I can get you a list of versions in a bit.

**Comment by ryanellis** [comment]
```Traceback (most recent call last):
  File "setup.py", line 30, in <module>
    'myapp = myapp:management_command',
  File "/usr/lib/python3.4/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.4/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.4/distutils/dist.py", line 974, in run_command
    cmd_obj.run()
  File "/usr/local/lib/python3.4/dist-packages/wheel/bdist_wheel.py", line 237, in run
    self.egg2dist(self.egginfo_dir, distinfo_dir)
  File "/usr/local/lib/python3.4/dist-packages/wheel/bdist_wheel.py", line 344, in egg2dist
    pkg_info = pkginfo_to_metadata(egginfo_path, pkginfo_path)
  File "/usr/local/lib/python3.4/dist-packages/wheel/metadata.py", line 88, in pkginfo_to_metadata
    for key, value in generate_requirements({extra: reqs}):
  File "/usr/local/lib/python3.4/dist-packages/wheel/metadata.py", line 67, in generate_requirements
    for new_req in convert_requirements(depends):
  File "/usr/local/lib/python3.4/dist-packages/wheel/metadata.py", line 36, in convert_requirements
    spec = requires_to_requires_dist(parsed_requirement)
  File "/usr/local/lib/python3.4/dist-packages/wheel/metadata.py", line 21, in requires_to_requires_dist
    if requirement.url:
AttributeError: 'Requirement' object has no attribute 'url'```

**Comment by benoit-pierre** [comment]
And the version of `pkg_resources` being used?

**Comment by onei** [comment]
Seeing the same error on CentOS7.4, thrown when using `python3 setup.py bdist_wheel`. Unfortunately it's not your standard CentOS install as it's one we've customised for internal use, but I don;t believe the Python packages are any different to those that come from the standard epel repos.

```
16:40:29 running install_scripts
16:40:29 Traceback (most recent call last):
16:40:29   File "setup.py", line 34, in <module>
16:40:29     'pycodestyle',
16:40:29   File "/usr/lib64/python3.4/distutils/core.py", line 148, in setup
16:40:29     dist.run_commands()
16:40:29   File "/usr/lib64/python3.4/distutils/dist.py", line 955, in run_commands
16:40:29     self.run_command(cmd)
16:40:29   File "/usr/lib64/python3.4/distutils/dist.py", line 974, in run_command
16:40:29     cmd_obj.run()
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/bdist_wheel.py", line 237, in run
16:40:29     self.egg2dist(self.egginfo_dir, distinfo_dir)
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/bdist_wheel.py", line 344, in egg2dist
16:40:29     pkg_info = pkginfo_to_metadata(egginfo_path, pkginfo_path)
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/metadata.py", line 88, in pkginfo_to_metadata
16:40:29     for key, value in generate_requirements({extra: reqs}):
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/metadata.py", line 67, in generate_requirements
16:40:29     for new_req in convert_requirements(depends):
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/metadata.py", line 36, in convert_requirements
16:40:29     spec = requires_to_requires_dist(parsed_requirement)
16:40:29   File "/usr/lib/python3.4/site-packages/wheel/metadata.py", line 21, in requires_to_requires_dist
16:40:29     if requirement.url:
16:40:29 AttributeError: 'Requirement' object has no attribute 'url'
```

**Comment by lyogev** [comment]
We're experiencing the same issue on machines in EMR (python 3.4) unable to build anything unless reverting to wheel 0.31.1 and below

**Comment by benoit-pierre** [comment]
Trusty is using pkg_resources/setuptools 3.3, which does use the packaging API, hence the issue (you need at least 18.8). I've got to ask, if you update wheel, why not update setuptools/pkg_resources too?

Anyway, please check the following patch fix your issue:
```diff
 wheel/metadata.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git i/wheel/metadata.py w/wheel/metadata.py
index ee74dd1..f566090 100644
--- i/wheel/metadata.py
+++ w/wheel/metadata.py
@@ -18,7 +18,7 @@
 
 def requires_to_requires_dist(requirement):
     """Return the version specifier for a requirement in PEP 345/566 fashion."""
-    if requirement.url:
+    if getattr(requirement, 'url', None):
         return " @ " + requirement.url
 
     requires_dist = []
```

**Comment by agronholm** [comment]
I was thinking of the same patch. I will do this once I get back from work. With this fix and the zero-permissions RECORD issue fixed I will then release wheel 0.32.1.

**Comment by agronholm** [comment]
FYI, I've released v0.32.1 which contains the fix for this.
